client/state/sites/reducer.js

```diff
@@ -144,7 +144,8 @@ export function items( state = {}, action ) {
 			// accounting for the fact that a non-existent icon property is
 			// equivalent to setting the media icon as null
 			const site = state[ siteId ];
-			if ( ! site || get( site.icon, 'media_id', null ) === mediaId ) {
+			if ( ! site || ( ! site.icon && null === mediaId ) ||
+					( site.icon && site.icon.media_id === mediaId ) ) {
```

Comment:
>Is this only for performance reasons? Maybe we should avoid this shortcut

Correct, it's to avoid updating sites state when the settings don't really affect any changes. Since the action isn't dispatched very frequently, it probably doesn't have a noticeable impact. If we run into more issues with the logic, I may be inclined to remove it, but since it's already pretty well tested here, I say it's worth keeping for now.